### PR TITLE
Session tracking

### PR DESCRIPTION
**Pull request by kattrali**

### CODE REVIEW

**Comment by kattrali** [comment]
✅

File: DependencyInjection/ClientFactory.php [hunk 1]
```diff
@@ -190,7 +205,9 @@ public function __construct(
         $env = null,
         $stage = null,
         array $stages = null,
-        array $filters = null
+        array $filters = null,
+        $sessions = null,
```
Comment (GrahamCampbell) [inline DependencyInjection/ClientFactory.php]: phpdoc needs updating

File: DependencyInjection/ClientFactory.php [hunk 2]
```diff
@@ -331,4 +354,39 @@ protected function setupPaths(Client $client, $strip, $project, $root)
             }
         }
     }
+
+    /**
+     * Setup session tracking.
+     *
+     * @param \Bugsnag\Client $client
+     * @param string|null     $endpoint
+     */
```
Comment (GrahamCampbell) [inline DependencyInjection/ClientFactory.php]: missing return void annotation

File: DependencyInjection/ClientFactory.php [hunk 3]
```diff
@@ -331,4 +354,39 @@ protected function setupPaths(Client $client, $strip, $project, $root)
...
+            $sessionTracker = $client->getSessionTracker();
+
+            $cache = new \Symfony\Component\Cache\Adapter\FilesystemAdapter();
+
+            $genericStorage = function ($key, $value = null) use ($cache) {
+                if (is_null($value)) {
+                    if ($cache->hasItem($key)) {
+                        return $cache->getItem($key)->get();
+                    } else {
+                        return;
```
Comment (GrahamCampbell) [inline DependencyInjection/ClientFactory.php]: is this needed?

Comment (GrahamCampbell) [reply DependencyInjection/ClientFactory.php]:
```php
if (is_null($value) && $cache->hasItem($key)) {
    return $cache->getItem($key)->get();
 } else {
    $item = $cache->getItem($key);
    $item->set($value);
    $cache->save($item);
}

Comment (GrahamCampbell) [reply DependencyInjection/ClientFactory.php]: Though, there is a race condition here anywhere, and two cache calls instead of one.

Comment (Cawllec) [reply DependencyInjection/ClientFactory.php]: Yeah, needs syncing of some sort.

**Comment by Cawllec** [comment]
I've added locking that should work for Symfony2.6+, but would like your opinion on it @GrahamCampbell to see if it's too heavy/there's an easier way.

**Comment by GrahamCampbell** [comment]
I'll take a look at this in the next 48 hours. :)

File: DependencyInjection/ClientFactory.php [hunk 4]
```diff
@@ -7,6 +7,11 @@
 use Bugsnag\Callbacks\CustomUser;
 use Bugsnag\Client;
 use Bugsnag\Configuration as Config;
+use Symfony\Component\Cache\Adapter\FilesystemAdapter;
```
Comment (GrahamCampbell) [inline DependencyInjection/ClientFactory.php]: Hmm, we probably don't want to force this on people. What if people are running on a system where they cannot write to the disk, not just for security, but for a serverless cloud deployment, etc.

File: DependencyInjection/ClientFactory.php (hunk 3 above)
Comment (GrahamCampbell) [reply DependencyInjection/ClientFactory.php]: Just do the get by itself, right?

File: DependencyInjection/ClientFactory.php (hunk 4 above)
Comment (Cawllec) [reply DependencyInjection/ClientFactory.php]: True.  I'm not sure how we'd proceed with storing session counts between sessions without it tho.  We could allow the user to specify the adapter, but without one we don't want to be firing off posts every time a request is made.

**Comment by DayS** [comment]
What's the status on this one ? Do you need some more hands to work on it ?

File: EventListener/BugsnagListener.php [hunk 5]
```diff
@@ -62,6 +62,10 @@ public function onKernelRequest(GetResponseEvent $event)
             return;
         }
 
+        if ($this->client->shouldCaptureSessions()) {
```
Comment (baumann-bestit) [inline EventListener/BugsnagListener.php:65]: When will this feature be implemented?

**Pull request closed without merging**